File: 2022/12_hill_climbing_algorithm/part1.py
from collections import defaultdict
from math import inf

def searchMap(old, x, y, count, path):
    if path[(x,y)]:
        return inf
    path[(x,y)] = True
    if x < 0 or x > len(map[0])-1 or y < 0 or y > len(map)-1:
        return inf
    new = map[y][x]
    if new == "E":
        if ord('z') - ord(old) > 1:
            return inf
        return count
    if ord(new) - ord(old) > 1:
        return inf
    else:
        return min(searchMap(new, x+1, y, count+1, path.copy()), searchMap(new, x-1, y, count+1, path.copy()), searchMap(new, x, y+1, count+1, path.copy()), searchMap(new, x, y-1, count+1, path.copy()))

map = []

Sx = -1
Sy = -1
Ex = -1
Ey = -1
with open("input.txt", "r") as file:
    for idx, line in enumerate(file):
        line = line.strip()
        row = []
        for jdx, c in enumerate(line):
            row.append(c)
            if c == "S":
                Sx = jdx
                Sy = idx
                row.pop()
                row.append('a')
            elif c == "E":
                Ex = jdx
                Ey = idx
        map.append(row)

print(searchMap("a", Sx, Sy, 0, defaultdict(bool)))

File: 2022/12_hill_climbing_algorithm/part1.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func getKey(x int, y int) string {
	s := strconv.Itoa(x)
	t := strconv.Itoa(y)
	return s + "_" + t
}

func searchMap(elevation [][]rune, old rune, x int, y int, count int, path map[string]bool) int {
	if path[getKey(x, y)] {
		fmt.Print("a")
		return math.MaxInt
	}
	path[getKey(x, y)] = true
	if x < 0 || x >= len(elevation[0]) || y < 0 || y >= len(elevation) {
		fmt.Print("b")
		return math.MaxInt
	}
	new := elevation[y][x]
	if new == 69 {
		if 'z'-old > 1 {
			fmt.Print("c")
			return math.MaxInt
		} else {
			return count
		}
	} else if new-old > 1 {
		fmt.Print("d")
		return math.MaxInt
	} else {
		min := math.MaxInt
		if x := searchMap(elevation, new, x+1, y, count+1, path); x < min {
			min = x
		}
		if x := searchMap(elevation, new, x-1, y, count+1, path); x < min {
			min = x
		}
		if x := searchMap(elevation, new, x, y+1, count+1, path); x < min {
			min = x
		}
		if x := searchMap(elevation, new, x, y-1, count+1, path); x < min {
			min = x
		}
		return min
	}
}

func main() {
	input, err := os.Open("test.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer input.Close()

	elevation := [][]rune{}

	Sx := -1
	Sy := -1
	y := 0
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		row := []rune{}
		for x, c := range line {
			if c == 'S' {
				row = append(row, 'a')
				Sx = x
				Sy = y
			} else {
				row = append(row, c)
			}
		}
		elevation = append(elevation, row)
		y++
	}

	res := searchMap(elevation, 'a', Sx, Sy, 0, map[string]bool{})
	if res == math.MaxInt {
		fmt.Println("Broken")
	} else {
		fmt.Println(res)
	}

	if err := scanner.Err(); err != nil {
		fmt.Println(err)
	}
}
